src/activity_handlers.cpp

```diff
@@ -481,417 +481,195 @@ int butcher_time_to_cut( const player &u, const item &corpse_item, const butcher
-                              !chunk.made_of( material_id( "veggy" ) );
-            if( action == F_DRESS ) {
-                for( int i = 1; i < pieces; ++i ) {
-                    if( make_offal && one_in( 3 ) ) {
-                        parts.charges++;
-                    }
-                }
-                chunk.charges = 0;
-            } else {
-                for( int i = 1; i < pieces; ++i ) {
-                    if( make_offal && one_in( 6 ) ) {
-                        parts.charges++;
-                    } else {
-                        chunk.charges++;
-                    }
-                }
+            if( entry_type == "flesh" ) {
+                return round( weight * damage );
             }
-            if( chunk.charges > 0 ) {
-                g->m.add_item_or_charges( p->pos(), chunk );
+            break;
+        case 4: // "pulped"
+            if( entry_type == "offal" ) {
+                return destroyed;
             }
-            if( parts.charges > 0 ) {
-                g->m.add_item_or_charges( p->pos(), parts );
+            if( entry_type == "skin" ) {
+                return destroyed;
             }
-        }
+            if( entry_type == "bone" ) {
+                return round( weight * damage );
+            }
+            if( entry_type == "flesh" ) {
+                return round( weight * high_damage );
+            }
+            break;
+        default: // "bruised" modifier is almost impossible to avoid; also includes no modifier (zero damage)
+            break;
     }
+    return weight;
 }
 
-void butchery_drops_harvest( item *corpse_item, const mtype &mt, player &p, const time_point &age,
-                             const std::function<int()> &roll_butchery, butcher_type action )
+void butchery_drops_harvest( item *corpse_item, const mtype &mt, player &p, const time_point &bday,
+                             const std::function<int()> &roll_butchery, butcher_type action,
+                             const std::function<double()> &roll_drops )
 {
     p.add_msg_if_player( m_neutral, _( mt.harvest->message().c_str() ) );
-
+    int monster_weight = to_gram( mt.weight );
+    monster_weight += round( monster_weight * rng_float( -0.1, 0.1 ) );
+    if( corpse_item->has_flag( "QUARTERED" ) ) {
+        monster_weight /= 4;
+    }
+    if( corpse_item->has_flag( "GIBBED" ) ) {
+        monster_weight = round( 0.85 * monster_weight );
+        if( action != F_DRESS ) {
+            p.add_msg_if_player( m_bad,
+                                 _( "You salvage what you can from the corpse, but it is badly damaged." ) );
+        }
+    }
+    int monster_weight_remaining = monster_weight;
     int practice = 4 + roll_butchery();
-    for( const auto &entry : *mt.harvest ) {
+
+    harvest_list harvest = *mt.harvest;
+    if( harvest.is_null() ) {
+        harvest = butchery_flags_deprecate( mt );
```

Comment:
Please check for a non-null harvest id and only dereference it when it's non-null. How about returning the *id* of the harvest list from this function?
```C++
harvest_id hid = mt.harvest.is_null() ? butchery_flags_deprecated( mt ) : mt.harvest;
const harvest_list &harvest = *hid; // no need to copy the list
```
This avoids copying the harvest list, looking it up via the `harvest::all().at` (which may throw).